aiida/engine/daemon/execmanager.py

```diff
@@ -75,55 +81,78 @@ def upload_calculation(node, transport, calc_info, folder, inputs=None, dry_run=
                     node.pk, computer.name))
 
-        # If it already exists, no exception is raised
-        try:
-            transport.chdir(remote_working_directory)
-        except IOError:
-            logger.debug(
-                '[submission of calculation {}] Unable to chdir in {}, trying to create it'.format(
-                    node.pk, remote_working_directory))
+        if calc_info.target_path:
+            pathlist = os.path.normpath(calc_info.target_path).split(os.sep)
+        else:
+            pathlist = [None]
+
+        # If it is an absolute path it will start checking dirs from the root
+        if pathlist[0] == '':
+            transport.chdir('/')
+            pathlist.pop(0)
+
+        # Else (relative path or no target path provided), it will start from
+        # the computer working directory
+        else:
+            # If it already exists, no exception is raised
             try:
-                transport.makedirs(remote_working_directory)
                 transport.chdir(remote_working_directory)
-            except EnvironmentError as exc:
-                raise exceptions.ConfigurationError(
-                    '[submission of calculation {}] '
-                    'Unable to create the remote directory {} on '
-                    "computer '{}': {}".format(
-                        node.pk, remote_working_directory, computer.name, exc))
-        # Store remotely with sharding (here is where we choose
-        # the folder structure of remote jobs; then I store this
-        # in the calculation properties using _set_remote_dir
-        # and I do not have to know the logic, but I just need to
-        # read the absolute path from the calculation properties.
-        transport.mkdir(calc_info.uuid[:2], ignore_existing=True)
-        transport.chdir(calc_info.uuid[:2])
-        transport.mkdir(calc_info.uuid[2:4], ignore_existing=True)
-        transport.chdir(calc_info.uuid[2:4])
+            except IOError:
+                logger.debug(
+                    '[submission of calculation {}] Unable to chdir in {}, trying to create it'.format(
+                        node.pk, remote_working_directory))
+                try:
+                    transport.makedirs(remote_working_directory)
+                    transport.chdir(remote_working_directory)
+                except EnvironmentError as exc:
+                    raise exceptions.ConfigurationError(
+                        '[submission of calculation {}] '
+                        'Unable to create the remote directory {} on '
+                        "computer '{}': {}".format(
+                            node.pk, remote_working_directory, computer.name, exc))
+
+        if calc_info.target_path:
+            for pathpart in pathlist:
+                try:
+                    transport.chdir(pathpart)
+                except IOError:
+                    transport.mkdir(pathpart)
+                    transport.chdir(pathpart)
```

Comment:
All of this logic should be simplified and improved a lot. All that this is doing is given an absolute path, try to create the path and chdir into it. This should be independent of whether `calc_info.target_path` is actually defined, so all this different logic can be gotten rid off

Reply:
Mmm ok, but the existing logic in the execmanager has hardcoded the logic of the 2:2:32 sharding, do you want me to fully replace this for something more versatile? If that is the case, do you think that the "path as list" approach is good? So I mean:

1. set remote_working_directory to the default sharding or to calc_info.target_path
2. ```pathlist = os.path.normpath(remote_working_directory).split(os.sep)```
3. Go element by element making sure the dir exists until the last one
4. Last one should be new or else apply existing procedure of moving this to REMOTE_WORK_DIRECTORY_LOST_FOUND

Reply:
I would determine the dirname, so the full path minus the last section, and use `os.makedirs(dirname, exist_ok=True)` which will not raise since [Python 3.4](https://docs.python.org/3/library/os.html#os.makedirs). Then you just create the leaf, which if it already exists, you enter the procedure for backing up. The only part that is different for your application is determination of `dirname`. So there you have a single switch and then all the rest becomes the same code

Reply:
This would have to be implemented **through the transport**, right? Because I can't directly use `os` to do stuff on the "remote" machine.